vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/customresourcevalidation/proxy/validate_proxy.go

```diff
@@ -0,0 +1,286 @@
+		if err := validateURI(spec.HTTPProxy); err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("HTTPProxy"), spec.HTTPProxy, err.Error()))
+		}
+	}
+	if len(spec.HTTPSProxy) != 0 {
+		if err := validateURI(spec.HTTPSProxy); err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("HTTPSProxy"), spec.HTTPSProxy, err.Error()))
+		}
+	}
+	if len(spec.ReadinessEndpoints) != 0 {
+		for _, endpoint := range spec.ReadinessEndpoints {
+			if err := validateURI(endpoint); err != nil {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("ReadinessEndpoints"), spec.ReadinessEndpoints, err.Error()))
+			} else {
+				if err := validateReadinessEndpoint(endpoint); err != nil {
+					allErrs = append(allErrs, field.Invalid(field.NewPath("spec.ReadinessEndpoints"), spec.ReadinessEndpoints, err.Error()))
+				}
+			}
+		}
+	}
+	if len(spec.NoProxy) != 0 {
+		for _, v := range strings.Split(spec.NoProxy, ",") {
+			v = strings.TrimSpace(v)
+			errDomain := validateDomainName(v, false)
+			_, _, errCIDR := net.ParseCIDR(v)
+			if errDomain != nil && errCIDR != nil {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("NoProxy"), v, "must be a CIDR or domain, without wildcard characters and without leading or trailing dots ('.')"))
+			}
+			noProxy, err := createNoProxy(installConfig, network)
+		}
+	}
+
+	return allErrs
+}
+
+func (proxyV1) ValidateCreate(uncastObj runtime.Object) field.ErrorList {
+	obj, allErrs := toProxyV1(uncastObj)
+	if len(allErrs) > 0 {
+		return allErrs
+	}
+
+	allErrs = append(allErrs, validateProxySpec(obj.Spec)...)
+
+	return allErrs
+}
+
+func (proxyV1) ValidateUpdate(uncastObj runtime.Object, uncastOldObj runtime.Object) field.ErrorList {
+	obj, allErrs := toProxyV1(uncastObj)
+	if len(allErrs) > 0 {
+		return allErrs
+	}
+	oldObj, allErrs := toProxyV1(uncastOldObj)
+	if len(allErrs) > 0 {
+		return allErrs
+	}
+
+	allErrs = append(allErrs, validation.ValidateObjectMetaUpdate(&obj.ObjectMeta, &oldObj.ObjectMeta, field.NewPath("metadata"))...)
+	allErrs = append(allErrs, validateProxySpec(obj.Spec)...)
+
+	return allErrs
+}
+
+// validateReadinessEndpoint validates a proxy readinessendpoint.
+func validateReadinessEndpoint(endpoint string) error {
+	err := validateReadinessEndpointWithRetries(endpoint, maxRetries)
+		if err != nil {
+			return err
+		}
+
+	return nil
+}
+
+// validateReadinessEndpointWithRetries tries to validate the proxy readinessendpoint in a finite loop,
+// it returns the last result if it never succeeds.
+func validateReadinessEndpointWithRetries(endpoint string, retries int) field.ErrorList {
+	var err error
+	for i := 0; i < retries; i++ {
+		result, output, err = runHTTPReadinessProbe(endpoint)
+		if err == nil {
+			return nil
+		}
+	}
+	return err
+}
+
+// runHTTPReadinessProbe issues an http GET request to endpoint and returns an error
+// if a 2XX or 3XX http status code is not returned.
+func runHTTPReadinessProbe(endpoint string) error {
```

Comment:
can we split the file into validate.go, admission.go and readiness.go or something like that?